File: win/MD3View/MD3Types.cs
using System.Runtime.InteropServices;

namespace MD3View;

// ============================================================
// On-disk MD3 binary format structures
// ============================================================

public static class MD3Constants
{
    public const int MD3_IDENT = ('3' << 24) + ('P' << 16) + ('D' << 8) + 'I';
    public const int MD3_VERSION = 15;
    public const float MD3_XYZ_SCALE = 1.0f / 64.0f;
    public const int MAX_QPATH = 64;

    public const int MD3_MAX_FRAMES = 1024;
    public const int MD3_MAX_TAGS = 16;
    public const int MD3_MAX_SURFACES = 32;
    public const int MD3_MAX_TRIANGLES = 8192;
    public const int MD3_MAX_VERTS = 4096;
    public const int MD3_MAX_SHADERS = 256;
}

[StructLayout(LayoutKind.Sequential, Pack = 1)]
public struct MD3DiskHeader
{
    public int Ident;
    public int Version;
    [MarshalAs(UnmanagedType.ByValArray, SizeConst = 64)]
    public byte[] Name;
    public int Flags;
    public int NumFrames;
    public int NumTags;
    public int NumSurfaces;
    public int NumSkins;
    public int OfsFrames;
    public int OfsTags;
    public int OfsSurfaces;
    public int OfsEnd;
}

[StructLayout(LayoutKind.Sequential, Pack = 1)]
public unsafe struct MD3DiskFrame
{
    public fixed float Bounds[6]; // [2][3]
    public fixed float LocalOrigin[3];
    public float Radius;
    public fixed byte Name[16];
}

[StructLayout(LayoutKind.Sequential, Pack = 1)]
public unsafe struct MD3DiskTag
{
    public fixed byte Name[64]; // MAX_QPATH
    public fixed float Origin[3];
    public fixed float Axis[9]; // [3][3]
}

[StructLayout(LayoutKind.Sequential, Pack = 1)]
public unsafe struct MD3DiskSurface
{
    public int Ident;
    public fixed byte Name[64]; // MAX_QPATH
    public int SurfFlags;
    public int NumFrames;
    public int NumShaders;
    public int NumVerts;
    public int NumTriangles;
    public int OfsTriangles;
    public int OfsShaders;
    public int OfsSt;
    public int OfsXyzNormals;
    public int OfsEnd;
}

[StructLayout(LayoutKind.Sequential, Pack = 1)]
public unsafe struct MD3DiskShader
{
    public fixed byte Name[64]; // MAX_QPATH
    public int ShaderIndex;
}

[StructLayout(LayoutKind.Sequential, Pack = 1)]
public struct MD3DiskTriangle
{
    public int Index0;
    public int Index1;
    public int Index2;
}

[StructLayout(LayoutKind.Sequential, Pack = 1)]
public struct MD3DiskTexCoord
{
    public float S;
    public float T;
}

[StructLayout(LayoutKind.Sequential, Pack = 1)]
public struct MD3DiskVertex
{
    public short X;
    public short Y;
    public short Z;
    public short Normal;
}

// ============================================================
// Runtime structures (decompressed, ready for rendering)
// ============================================================

public struct MD3Vertex
{
    public float PosX, PosY, PosZ;
    public float NormX, NormY, NormZ;
}

public class MD3Surface
{
    public string Name = "";
    public int NumFrames;
    public int NumVerts;
    public int NumTriangles;
    public int[] Triangles = Array.Empty<int>();     // numTriangles * 3
    public float[] TexCoords = Array.Empty<float>(); // numVerts * 2
    public MD3Vertex[] Vertices = Array.Empty<MD3Vertex>(); // numVerts * numFrames
    public string ShaderName = "";
    public int TextureID; // GL texture ID
}

public struct MD3Tag
{
    public string Name;
    public float[] Origin;    // [3]
    public float[,] Axis;    // [3,3]

    public MD3Tag()
    {
        Name = "";
        Origin = new float[3];
        Axis = new float[3, 3];
    }
}

public struct MD3Frame
{
    public float[] Bounds;       // [6] = [2][3]
    public float[] LocalOrigin;  // [3]
    public float Radius;
    public string Name;
}

// ============================================================
// Animation enum (matches Q3 bg_public.h)
// ============================================================

public enum AnimNumber
{
    BOTH_DEATH1,
    BOTH_DEAD1,
    BOTH_DEATH2,
    BOTH_DEAD2,
    BOTH_DEATH3,
    BOTH_DEAD3,

    TORSO_GESTURE,

    TORSO_ATTACK,
    TORSO_ATTACK2,

    TORSO_DROP,
    TORSO_RAISE,

    TORSO_STAND,
    TORSO_STAND2,

    LEGS_WALKCR,
    LEGS_WALK,
    LEGS_RUN,
    LEGS_BACK,
    LEGS_SWIM,

    LEGS_JUMP,
    LEGS_LAND,

    LEGS_JUMPB,
    LEGS_LANDB,

    LEGS_IDLE,
    LEGS_IDLECR,

    LEGS_TURN,

    TORSO_GETFLAG,
    TORSO_GUARDBASE,
    TORSO_PATROL,
    TORSO_FOLLOWME,
    TORSO_AFFIRMATIVE,
    TORSO_NEGATIVE,

    MAX_ANIMATIONS,

    LEGS_BACKCR,
    LEGS_BACKWALK,
    FLAG_RUN,
    FLAG_STAND,
    FLAG_STAND2RUN,

    MAX_TOTALANIMATIONS
}

public static class AnimationNames
{
    public static readonly string[] Names =
    {
        "BOTH_DEATH1",
        "BOTH_DEAD1",
        "BOTH_DEATH2",
        "BOTH_DEAD2",
        "BOTH_DEATH3",
        "BOTH_DEAD3",
        "TORSO_GESTURE",
        "TORSO_ATTACK",
        "TORSO_ATTACK2",
        "TORSO_DROP",
        "TORSO_RAISE",
        "TORSO_STAND",
        "TORSO_STAND2",
        "LEGS_WALKCR",
        "LEGS_WALK",
        "LEGS_RUN",
        "LEGS_BACK",
        "LEGS_SWIM",
        "LEGS_JUMP",
        "LEGS_LAND",
        "LEGS_JUMPB",
        "LEGS_LANDB",
        "LEGS_IDLE",
        "LEGS_IDLECR",
        "LEGS_TURN",
        "TORSO_GETFLAG",
        "TORSO_GUARDBASE",
        "TORSO_PATROL",
        "TORSO_FOLLOWME",
        "TORSO_AFFIRMATIVE",
        "TORSO_NEGATIVE",
    };
}

public struct Animation
{
    public int FirstFrame;
    public int NumFrames;
    public int LoopFrames;
    public int FrameLerp; // msec between frames
    public int Reversed;
    public int Flipflop;
}

public struct AnimState
{
    public AnimNumber AnimIndex;
    public int CurrentFrame;
    public int NextFrame;
    public float Fraction;    // lerp fraction between frames
    public double FrameTime;  // time of last frame change (ms)
    public bool Playing;
}

public struct TagTransform
{
    public float[] Origin;    // [3]
    public float[,] Axis;    // [3,3]

    public TagTransform()
    {
        Origin = new float[3];
        Axis = new float[3, 3];
    }

    public static TagTransform Identity()
    {
        var t = new TagTransform();
        t.Axis[0, 0] = 1;
        t.Axis[1, 1] = 1;
        t.Axis[2, 2] = 1;
        return t;
    }
}

File: lin/md3_types.py
"""MD3 binary format types, animation enums, and runtime dataclasses."""

import struct
from dataclasses import dataclass, field
from enum import IntEnum
from typing import List

# On-disk MD3 constants
MD3_IDENT = (ord('3') << 24) + (ord('P') << 16) + (ord('D') << 8) + ord('I')
MD3_VERSION = 15
MD3_XYZ_SCALE = 1.0 / 64.0
MAX_QPATH = 64

# MD3 limits
MD3_MAX_FRAMES = 1024
MD3_MAX_TAGS = 16
MD3_MAX_SURFACES = 32
MD3_MAX_TRIANGLES = 8192
MD3_MAX_VERTS = 4096
MD3_MAX_SHADERS = 256

# Struct format strings for on-disk types (little-endian)
# MD3DiskHeader: ident, version, name[64], flags, numFrames, numTags, numSurfaces,
#                numSkins, ofsFrames, ofsTags, ofsSurfaces, ofsEnd
MD3_DISK_HEADER_FMT = '<ii64siiiiiiiii'
MD3_DISK_HEADER_SIZE = struct.calcsize(MD3_DISK_HEADER_FMT)

# MD3DiskFrame: bounds[2][3], localOrigin[3], radius, name[16]
MD3_DISK_FRAME_FMT = '<6f3ff16s'
MD3_DISK_FRAME_SIZE = struct.calcsize(MD3_DISK_FRAME_FMT)

# MD3DiskTag: name[64], origin[3], axis[3][3]
MD3_DISK_TAG_FMT = '<64s3f9f'
MD3_DISK_TAG_SIZE = struct.calcsize(MD3_DISK_TAG_FMT)

# MD3DiskSurface: ident, name[64], flags, numFrames, numShaders, numVerts,
#                 numTriangles, ofsTriangles, ofsShaders, ofsSt, ofsXyzNormals, ofsEnd
MD3_DISK_SURFACE_FMT = '<i64siiiiiiiiii'
MD3_DISK_SURFACE_SIZE = struct.calcsize(MD3_DISK_SURFACE_FMT)

# MD3DiskShader: name[64], shaderIndex
MD3_DISK_SHADER_FMT = '<64si'
MD3_DISK_SHADER_SIZE = struct.calcsize(MD3_DISK_SHADER_FMT)

# MD3DiskTriangle: indexes[3]
MD3_DISK_TRIANGLE_FMT = '<3i'
MD3_DISK_TRIANGLE_SIZE = struct.calcsize(MD3_DISK_TRIANGLE_FMT)

# MD3DiskTexCoord: st[2]
MD3_DISK_TEXCOORD_FMT = '<2f'
MD3_DISK_TEXCOORD_SIZE = struct.calcsize(MD3_DISK_TEXCOORD_FMT)

# MD3DiskVertex: xyz[3] as int16, normal as int16
MD3_DISK_VERTEX_FMT = '<3hh'
MD3_DISK_VERTEX_SIZE = struct.calcsize(MD3_DISK_VERTEX_FMT)


# Animation enum (matches Q3 bg_public.h)
class AnimNumber(IntEnum):
    BOTH_DEATH1 = 0
    BOTH_DEAD1 = 1
    BOTH_DEATH2 = 2
    BOTH_DEAD2 = 3
    BOTH_DEATH3 = 4
    BOTH_DEAD3 = 5

    TORSO_GESTURE = 6

    TORSO_ATTACK = 7
    TORSO_ATTACK2 = 8

    TORSO_DROP = 9
    TORSO_RAISE = 10

    TORSO_STAND = 11
    TORSO_STAND2 = 12

    LEGS_WALKCR = 13
    LEGS_WALK = 14
    LEGS_RUN = 15
    LEGS_BACK = 16
    LEGS_SWIM = 17

    LEGS_JUMP = 18
    LEGS_LAND = 19

    LEGS_JUMPB = 20
    LEGS_LANDB = 21

    LEGS_IDLE = 22
    LEGS_IDLECR = 23

    LEGS_TURN = 24

    TORSO_GETFLAG = 25
    TORSO_GUARDBASE = 26
    TORSO_PATROL = 27
    TORSO_FOLLOWME = 28
    TORSO_AFFIRMATIVE = 29
    TORSO_NEGATIVE = 30

    MAX_ANIMATIONS = 31

    LEGS_BACKCR = 32
    LEGS_BACKWALK = 33
    FLAG_RUN = 34
    FLAG_STAND = 35
    FLAG_STAND2RUN = 36

    MAX_TOTALANIMATIONS = 37


ANIMATION_NAMES = [
    "BOTH_DEATH1", "BOTH_DEAD1", "BOTH_DEATH2", "BOTH_DEAD2",
    "BOTH_DEATH3", "BOTH_DEAD3",
    "TORSO_GESTURE",
    "TORSO_ATTACK", "TORSO_ATTACK2",
    "TORSO_DROP", "TORSO_RAISE",
    "TORSO_STAND", "TORSO_STAND2",
    "LEGS_WALKCR", "LEGS_WALK", "LEGS_RUN", "LEGS_BACK", "LEGS_SWIM",
    "LEGS_JUMP", "LEGS_LAND",
    "LEGS_JUMPB", "LEGS_LANDB",
    "LEGS_IDLE", "LEGS_IDLECR",
    "LEGS_TURN",
    "TORSO_GETFLAG", "TORSO_GUARDBASE", "TORSO_PATROL",
    "TORSO_FOLLOWME", "TORSO_AFFIRMATIVE", "TORSO_NEGATIVE",
]


@dataclass
class Animation:
    firstFrame: int = 0
    numFrames: int = 0
    loopFrames: int = 0
    frameLerp: int = 0
    reversed: int = 0
    flipflop: int = 0


@dataclass
class AnimState:
    animIndex: int = 0
    currentFrame: int = 0
    nextFrame: int = 0
    fraction: float = 0.0
    frameTime: float = 0.0
    playing: bool = True


@dataclass
class TagTransform:
    origin: List[float] = field(default_factory=lambda: [0.0, 0.0, 0.0])
    axis: List[List[float]] = field(default_factory=lambda: [
        [1.0, 0.0, 0.0],
        [0.0, 1.0, 0.0],
        [0.0, 0.0, 1.0],
    ])


# Runtime structures
@dataclass
class MD3Vertex:
    position: List[float] = field(default_factory=lambda: [0.0, 0.0, 0.0])
    normal: List[float] = field(default_factory=lambda: [0.0, 0.0, 0.0])


@dataclass
class MD3Tag:
    name: str = ""
    origin: List[float] = field(default_factory=lambda: [0.0, 0.0, 0.0])
    axis: List[List[float]] = field(default_factory=lambda: [
        [1.0, 0.0, 0.0],
        [0.0, 1.0, 0.0],
        [0.0, 0.0, 1.0],
    ])


@dataclass
class MD3Frame:
    bounds: List[List[float]] = field(default_factory=lambda: [[0.0]*3, [0.0]*3])
    localOrigin: List[float] = field(default_factory=lambda: [0.0, 0.0, 0.0])
    radius: float = 0.0
    name: str = ""


@dataclass
class MD3Surface:
    name: str = ""
    numFrames: int = 0
    numVerts: int = 0
    numTriangles: int = 0
    triangles: List[int] = field(default_factory=list)      # flat: numTriangles * 3
    texCoords: List[float] = field(default_factory=list)     # flat: numVerts * 2
    vertices: List[MD3Vertex] = field(default_factory=list)  # numVerts * numFrames
    shaderName: str = ""
    textureID: int = 0
